File: plugins/agentic-sdlc/src/agentic_sdlc/git/worktree.py
"""Git worktree management for parallel workflow execution."""

from __future__ import annotations

import secrets
import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


def get_executable(name: str) -> str:
    """Resolve executable path for cross-platform subprocess calls.

    Uses shutil.which() to find the full path, allowing shell=False
    in subprocess calls while maintaining Windows compatibility.

    Args:
        name: Executable name (e.g., "claude", "git")

    Returns:
        Full path to the executable

    Raises:
        FileNotFoundError: If executable not found in PATH
    """
    path = shutil.which(name)
    if not path:
        raise FileNotFoundError(f"Executable not found in PATH: {name}")
    return path


@dataclass
class Worktree:
    """Git worktree information."""

    path: Path
    branch: str
    base_branch: str

    def exists(self) -> bool:
        """Check if the worktree directory exists."""
        return self.path.exists()


def _run_git(
    args: list[str],
    cwd: Path | None = None,
    check: bool = True,
) -> subprocess.CompletedProcess[str]:
    """Run a git command."""
    git_path = get_executable("git")
    cmd = [git_path] + args
    result = subprocess.run(
        cmd,
        capture_output=True,
        text=True,
        cwd=str(cwd) if cwd else None,
        shell=False,
    )
    if check and result.returncode != 0:
        raise RuntimeError(f"Git command failed: {' '.join(cmd)}\n{result.stderr}")
    return result


def _truncate(name: str, max_len: int = 30) -> str:
    """Truncate name to max length for Windows path safety."""
    return name[:max_len] if len(name) > max_len else name


def _generate_suffix() -> str:
    """Generate 6 character random suffix."""
    return secrets.token_hex(3)


def _sanitize_name(name: str) -> str:
    """Sanitize a name for use in file paths and branch names."""
    return name.replace("/", "-").replace(" ", "-").replace("_", "-").lower()


def get_repo_root(cwd: Path | None = None) -> Path:
    """Get the root directory of the git repository."""
    result = _run_git(["rev-parse", "--show-toplevel"], cwd=cwd)
    return Path(result.stdout.strip())


def get_default_branch(cwd: Path | None = None) -> str:
    """Get the default branch name for the repository."""
    result = _run_git(
        ["symbolic-ref", "refs/remotes/origin/HEAD"],
        cwd=cwd,
        check=False,
    )
    if result.returncode == 0:
        return result.stdout.strip().split("/")[-1]
    for branch in ["main", "master"]:
        result = _run_git(["rev-parse", "--verify", branch], cwd=cwd, check=False)
        if result.returncode == 0:
            return branch
    return "main"


def get_current_branch(cwd: Path | None = None) -> str:
    """Get the current branch name."""
    result = _run_git(["rev-parse", "--abbrev-ref", "HEAD"], cwd=cwd)
    return result.stdout.strip()


def create_worktree(
    workflow_name: str,
    step_name: str,
    base_branch: str | None = None,
    repo_root: Path | None = None,
) -> Worktree:
    """Create a new worktree for a workflow step.

    Naming convention:
    - Path: .worktrees/agentic-{workflow}-{step}-{random}
    - Branch: agentic/{workflow}-{step}-{random}
    """
    if repo_root is None:
        repo_root = get_repo_root()

    if base_branch is None:
        base_branch = get_default_branch(repo_root)

    suffix = _generate_suffix()
    wf_name = _truncate(_sanitize_name(workflow_name))
    st_name = _truncate(_sanitize_name(step_name))

    dir_name = f"agentic-{wf_name}-{st_name}-{suffix}"
    branch_name = f"agentic/{wf_name}-{st_name}-{suffix}"

    worktree_path = repo_root / ".worktrees" / dir_name
    worktree_path.parent.mkdir(parents=True, exist_ok=True)

    if worktree_path.exists():
        shutil.rmtree(worktree_path, ignore_errors=True)
        _run_git(["worktree", "prune"], cwd=repo_root, check=False)

    _run_git(
        ["worktree", "add", "-b", branch_name, str(worktree_path), base_branch],
        cwd=repo_root,
    )

    return Worktree(path=worktree_path, branch=branch_name, base_branch=base_branch)


def remove_worktree(
    worktree: Worktree,
    repo_root: Path | None = None,
    delete_branch: bool = True,
) -> None:
    """Remove a worktree and optionally delete the branch."""
    if repo_root is None:
        repo_root = get_repo_root()

    result = _run_git(
        ["worktree", "remove", "--force", str(worktree.path)],
        cwd=repo_root,
        check=False,
    )

    if result.returncode != 0 and worktree.path.exists():
        shutil.rmtree(worktree.path, ignore_errors=True)
        _run_git(["worktree", "prune"], cwd=repo_root, check=False)

    if delete_branch and worktree.branch:
        _run_git(["branch", "-D", worktree.branch], cwd=repo_root, check=False)


def list_worktrees(repo_root: Path | None = None) -> list[Worktree]:
    """List all worktrees in the repository."""
    if repo_root is None:
        repo_root = get_repo_root()

    result = _run_git(["worktree", "list", "--porcelain"], cwd=repo_root)

    worktrees = []
    current_path: Path | None = None
    current_branch = ""

    for line in result.stdout.strip().split("\n"):
        if line.startswith("worktree "):
            current_path = Path(line[9:])
        elif line.startswith("branch "):
            current_branch = line.replace("branch refs/heads/", "")
        elif line == "" and current_path:
            worktrees.append(Worktree(path=current_path, branch=current_branch, base_branch=""))
            current_path = None
            current_branch = ""

    if current_path:
        worktrees.append(Worktree(path=current_path, branch=current_branch, base_branch=""))

    return worktrees


def list_agentic_worktrees(repo_root: Path | None = None) -> list[Worktree]:
    """List only agentic worktrees."""
    worktrees = list_worktrees(repo_root)
    return [wt for wt in worktrees if wt.branch.startswith("agentic/")]


def prune_orphaned(repo_root: Path | None = None) -> int:
    """Prune orphaned worktrees and stale agentic worktrees.

    Returns number of worktrees cleaned up.
    """
    if repo_root is None:
        repo_root = get_repo_root()

    _run_git(["worktree", "prune"], cwd=repo_root, check=False)

    cleaned = 0
    worktrees_dir = repo_root / ".worktrees"
    if worktrees_dir.exists():
        for wt_dir in worktrees_dir.iterdir():
            if wt_dir.is_dir() and wt_dir.name.startswith("agentic-"):
                git_file = wt_dir / ".git"
                if not git_file.exists():
                    shutil.rmtree(wt_dir, ignore_errors=True)
                    cleaned += 1

    return cleaned


def create_branch(
    branch_name: str,
    base_branch: str | None = None,
    cwd: Path | None = None,
) -> str:
    """Create a new branch from the specified base."""
    if base_branch is None:
        base_branch = get_default_branch(cwd)

    _run_git(["checkout", "-b", branch_name, base_branch], cwd=cwd)
    return branch_name


def checkout_branch(branch_name: str, cwd: Path | None = None) -> None:
    """Checkout an existing branch."""
    _run_git(["checkout", branch_name], cwd=cwd)


def commit_changes(
    message: str,
    cwd: Path | None = None,
    add_all: bool = True,
) -> bool:
    """Commit staged changes.

    Returns True if a commit was made, False if nothing to commit.
    """
    if add_all:
        _run_git(["add", "-A"], cwd=cwd)

    result = _run_git(["status", "--porcelain"], cwd=cwd)
    if not result.stdout.strip():
        return False

    _run_git(["commit", "-m", message], cwd=cwd)
    return True


def push_branch(
    branch_name: str,
    remote: str = "origin",
    cwd: Path | None = None,
    set_upstream: bool = True,
) -> None:
    """Push a branch to remote."""
    args = ["push"]
    if set_upstream:
        args.extend(["-u", remote, branch_name])
    else:
        args.extend([remote, branch_name])
    _run_git(args, cwd=cwd)

File: src/git/worktree.ts
/** Git worktree management for parallel workflow execution. */

import { execFileSync } from "node:child_process";
import { randomBytes } from "node:crypto";
import {
	existsSync,
	mkdirSync,
	readFileSync,
	readdirSync,
	rmSync,
	statSync,
	writeFileSync,
} from "node:fs";
import path from "node:path";
import { getExecutable } from "../runtimes/utils.js";
import type { WorktreeCleanup, WorktreeLocation } from "../types.js";

// --- Worktree data ---

export interface Worktree {
	path: string;
	branch: string;
	baseBranch: string;
}

// --- Git command helper ---

export function runGit(
	args: string[],
	cwd?: string | null,
	check = true,
): { stdout: string; stderr: string; returncode: number } {
	const gitPath = getExecutable("git");
	const cmd = [gitPath, ...args];
	try {
		const stdout = execFileSync(gitPath, args, {
			encoding: "utf-8",
			cwd: cwd ?? undefined,
			stdio: ["pipe", "pipe", "pipe"],
		});
		return { stdout, stderr: "", returncode: 0 };
	} catch (e: unknown) {
		const err = e as { stdout?: string; stderr?: string; status?: number };
		if (check) {
			throw new Error(`Git command failed: ${cmd.join(" ")}\n${err.stderr ?? ""}`);
		}
		return {
			stdout: (err.stdout as string) ?? "",
			stderr: (err.stderr as string) ?? "",
			returncode: (err.status as number) ?? 1,
		};
	}
}

// --- Name helpers ---

function truncate(name: string, maxLen = 30): string {
	return name.length > maxLen ? name.slice(0, maxLen) : name;
}

function generateSuffix(): string {
	return randomBytes(3).toString("hex");
}

function sanitizeName(name: string): string {
	return name.replace(/\//g, "-").replace(/ /g, "-").replace(/_/g, "-").toLowerCase();
}

// --- Repository helpers ---

export function getRepoRoot(cwd?: string | null): string {
	const result = runGit(["rev-parse", "--show-toplevel"], cwd);
	return result.stdout.trim();
}

export function getDefaultBranch(cwd?: string | null): string {
	const result = runGit(["symbolic-ref", "refs/remotes/origin/HEAD"], cwd, false);
	if (result.returncode === 0) {
		const parts = result.stdout.trim().split("/");
		return parts[parts.length - 1] ?? "main";
	}
	for (const branch of ["main", "master"]) {
		const check = runGit(["rev-parse", "--verify", branch], cwd, false);
		if (check.returncode === 0) {
			return branch;
		}
	}
	return "main";
}

export function getCurrentBranch(cwd?: string | null): string {
	const result = runGit(["rev-parse", "--abbrev-ref", "HEAD"], cwd);
	return result.stdout.trim();
}

// --- Worktree path resolution ---

function resolveWorktreePath(
	repoRoot: string,
	dirName: string,
	location: WorktreeLocation,
	directory: string | null,
): string {
	switch (location) {
		case "sibling":
			return path.join(path.dirname(repoRoot), ".worktrees", dirName);
		case "nested":
			return path.join(repoRoot, ".worktrees", dirName);
		case "absolute": {
			if (!directory) {
				throw new Error("Worktree location 'absolute' requires a 'directory' to be specified");
			}
			return path.join(directory, dirName);
		}
	}
}

function buildDirName(
	repoRoot: string,
	wfName: string,
	stName: string,
	suffix: string,
	location: WorktreeLocation,
): string {
	if (location === "nested") {
		return `agentic-${wfName}-${stName}-${suffix}`;
	}
	const repoName = sanitizeName(path.basename(repoRoot));
	return `${repoName}-${wfName}-${stName}-${suffix}`;
}

// --- .gitignore helper ---

function ensureGitignoreEntry(repoRoot: string, entry: string): void {
	const gitignorePath = path.join(repoRoot, ".gitignore");
	let content = "";
	if (existsSync(gitignorePath)) {
		content = readFileSync(gitignorePath, "utf-8");
	}
	const lines = content.split("\n");
	if (!lines.some((line) => line.trim() === entry)) {
		const updated = content.endsWith("\n") ? `${content}${entry}\n` : `${content}\n${entry}\n`;
		writeFileSync(gitignorePath, updated, "utf-8");
	}
}

// --- Worktree operations ---

export interface CreateWorktreeOptions {
	workflowName: string;
	stepName: string;
	/** When provided, used as the worktree directory and branch name instead of
	 *  generating from workflowName + stepName + random suffix. */
	workflowId?: string | null;
	baseBranch?: string | null;
	repoRoot?: string | null;
	location: WorktreeLocation;
	directory?: string | null;
}

/** Resolve worktree path with incremental suffix when the path already exists. */
function resolveWorktreePath_incremental(
	repoRoot: string,
	baseDirName: string,
	location: WorktreeLocation,
	directory: string | null,
): string {
	let worktreePath = resolveWorktreePath(repoRoot, baseDirName, location, directory);
	if (!existsSync(worktreePath)) {
		return worktreePath;
	}
	let counter = 2;
	while (existsSync(resolveWorktreePath(repoRoot, `${baseDirName}-${counter}`, location, directory))) {
		counter++;
	}
	return resolveWorktreePath(repoRoot, `${baseDirName}-${counter}`, location, directory);
}

export function createWorktree(options: CreateWorktreeOptions): Worktree {
	const {
		workflowName,
		stepName,
		workflowId,
		baseBranch,
		repoRoot: repoRootOpt,
		location,
		directory = null,
	} = options;

	const root = repoRootOpt || getRepoRoot();
	const base = baseBranch || getDefaultBranch(root);

	let dirName: string;
	let branchName: string;

	if (workflowId) {
		// Workflow-level worktree: use workflowId for consistent naming with output dir
		const safeId = sanitizeName(workflowId);
		if (location === "nested") {
			dirName = safeId;
		} else {
			const repoName = sanitizeName(path.basename(root));
			dirName = `${repoName}-${safeId}`;
		}
		branchName = `agentic/${safeId}`;
	} else {
		// Step-level worktree: use workflowName + stepName + random suffix
		const suffix = generateSuffix();
		const wfName = truncate(sanitizeName(workflowName));
		const stName = truncate(sanitizeName(stepName));
		dirName = buildDirName(root, wfName, stName, suffix, location);
		branchName = `agentic/${wfName}-${stName}-${suffix}`;
	}

	const worktreePath = resolveWorktreePath_incremental(root, dirName, location, directory);
	const parentDir = path.dirname(worktreePath);
	mkdirSync(parentDir, { recursive: true });

	// Derive the actual dirName for branch naming if incremented
	const actualDirName = path.basename(worktreePath);
	if (actualDirName !== dirName) {
		branchName = `agentic/${sanitizeName(actualDirName)}`;
	}

	// Add .worktrees/ to .gitignore for nested mode only
	if (location === "nested") {
		ensureGitignoreEntry(root, ".worktrees/");
	}

	runGit(["worktree", "add", "-b", branchName, worktreePath, base], root);

	return { path: worktreePath, branch: branchName, baseBranch: base };
}

export function removeWorktree(
	worktree: Worktree,
	repoRoot?: string | null,
	deleteBranch = true,
): void {
	const root = repoRoot || getRepoRoot();

	const result = runGit(["worktree", "remove", "--force", worktree.path], root, false);

	if (result.returncode !== 0 && existsSync(worktree.path)) {
		rmSync(worktree.path, { recursive: true, force: true });
		runGit(["worktree", "prune"], root, false);
	}

	if (deleteBranch && worktree.branch) {
		runGit(["branch", "-D", worktree.branch], root, false);
	}
}

// --- Safety commit ---

export interface SafetyCommitResult {
	committed: boolean;
	failed: boolean;
	error?: string;
}

/**
 * Commits uncommitted changes in a worktree before cleanup to prevent data loss.
 * Returns whether a commit was made, and whether the commit failed.
 * If the commit fails, callers should preserve the worktree rather than deleting it.
 */
export function safetyCommit(
	worktreePath: string,
	logger?: {
		info: (step: string, message: string) => void;
		warning: (step: string, message: string) => void;
	},
): SafetyCommitResult {
	const status = runGit(["status", "--porcelain"], worktreePath, false);
	if (!status.stdout.trim()) {
		return { committed: false, failed: false };
	}

	try {
		runGit(["add", "-A"], worktreePath);
		runGit(
			["commit", "-m", "chore: auto-save uncommitted changes before worktree cleanup"],
			worktreePath,
		);
		logger?.info("worktree", `Safety commit created in worktree: ${worktreePath}`);
		return { committed: true, failed: false };
	} catch (e: unknown) {
		const error = e instanceof Error ? e.message : String(e);
		logger?.warning(
			"worktree",
			`Safety commit failed in worktree ${worktreePath}: ${error}. Preserving worktree to prevent data loss.`,
		);
		return { committed: false, failed: true, error };
	}
}

// --- Pruning ---

export interface PruneLocation {
	location: WorktreeLocation;
	repoRoot: string;
	directory?: string | null;
}

export function listWorktrees(repoRoot?: string | null): Worktree[] {
	const root = repoRoot || getRepoRoot();

	const result = runGit(["worktree", "list", "--porcelain"], root);

	const worktrees: Worktree[] = [];
	let currentPath: string | null = null;
	let currentBranch = "";

	for (const line of result.stdout.trim().split("\n")) {
		if (line.startsWith("worktree ")) {
			currentPath = line.slice(9);
		} else if (line.startsWith("branch ")) {
			currentBranch = line.replace("branch refs/heads/", "");
		} else if (line === "" && currentPath) {
			worktrees.push({ path: currentPath, branch: currentBranch, baseBranch: "" });
			currentPath = null;
			currentBranch = "";
		}
	}

	if (currentPath) {
		worktrees.push({ path: currentPath, branch: currentBranch, baseBranch: "" });
	}

	return worktrees;
}

export function listAgenticWorktrees(repoRoot?: string | null): Worktree[] {
	return listWorktrees(repoRoot).filter((wt) => wt.branch.startsWith("agentic/"));
}

function pruneWorktreesDir(worktreesDir: string, prefix: string): number {
	let cleaned = 0;
	if (existsSync(worktreesDir)) {
		for (const entry of readdirSync(worktreesDir)) {
			if (!entry.startsWith(prefix)) continue;
			const wtDir = path.join(worktreesDir, entry);
			if (!statSync(wtDir).isDirectory()) continue;
			const gitFile = path.join(wtDir, ".git");
			if (!existsSync(gitFile)) {
				rmSync(wtDir, { recursive: true, force: true });
				cleaned++;
			}
		}
	}
	return cleaned;
}

export function pruneOrphaned(repoRoot?: string | null, prune?: PruneLocation): number {
	const root = repoRoot || getRepoRoot();

	runGit(["worktree", "prune"], root, false);

	if (prune) {
		const { location, directory } = prune;
		const pRoot = prune.repoRoot;
		let worktreesDir: string;
		let prefix: string;

		if (location === "nested") {
			worktreesDir = path.join(pRoot, ".worktrees");
			prefix = "agentic-";
		} else if (location === "sibling") {
			worktreesDir = path.join(path.dirname(pRoot), ".worktrees");
			const repoName = sanitizeName(path.basename(pRoot));
			prefix = `${repoName}-`;
		} else if (location === "absolute" && directory) {
			worktreesDir = directory;
			const repoName = sanitizeName(path.basename(pRoot));
			prefix = `${repoName}-`;
		} else {
			// Fallback to nested scan
			worktreesDir = path.join(pRoot, ".worktrees");
			prefix = "agentic-";
		}

		return pruneWorktreesDir(worktreesDir, prefix);
	}

	// Default: scan both nested (backward compat) and sibling (new default) locations
	let cleaned = 0;
	cleaned += pruneWorktreesDir(path.join(root, ".worktrees"), "agentic-");
	const siblingWorktreesDir = path.join(path.dirname(root), ".worktrees");
	const repoName = sanitizeName(path.basename(root));
	cleaned += pruneWorktreesDir(siblingWorktreesDir, `${repoName}-`);

	return cleaned;
}

// --- Branch operations ---

export function createBranch(
	branchName: string,
	baseBranch?: string | null,
	cwd?: string | null,
): string {
	const base = baseBranch || getDefaultBranch(cwd);
	runGit(["checkout", "-b", branchName, base], cwd);
	return branchName;
}

export function checkoutBranch(branchName: string, cwd?: string | null): void {
	runGit(["checkout", branchName], cwd);
}

export function commitChanges(message: string, cwd?: string | null, addAll = true): boolean {
	if (addAll) {
		runGit(["add", "-A"], cwd);
	}

	const result = runGit(["status", "--porcelain"], cwd);
	if (!result.stdout.trim()) {
		return false;
	}

	runGit(["commit", "-m", message], cwd);
	return true;
}

export function pushBranch(
	branchName: string,
	remote = "origin",
	cwd?: string | null,
	setUpstream = true,
): void {
	const args = ["push"];
	if (setUpstream) {
		args.push("-u", remote, branchName);
	} else {
		args.push(remote, branchName);
	}
	runGit(args, cwd);
}
